ext/liquid_c/vm.c

```diff
@@ -11,53 +12,394 @@ ID id_ivar_last_capture_length;
+static VALUE cLiquidCVM;
+
+typedef struct vm {
+    c_buffer_t stack;
+    VALUE strainer;
+    VALUE filter_methods;
+    VALUE interrupts;
+    VALUE resource_limits;
+    VALUE global_filter;
+    bool strict_filters;
+} vm_t;
+
+static void vm_mark(void *ptr)
 {
-    Check_Type(output, T_STRING);
+    vm_t *vm = ptr;
+    rb_gc_mark_locations((VALUE *)vm->stack.data, (VALUE *)vm->stack.data_end);
+    rb_gc_mark(vm->strainer);
+    rb_gc_mark(vm->filter_methods);
+    rb_gc_mark(vm->interrupts);
+    rb_gc_mark(vm->resource_limits);
+}
 
-    VALUE resource_limits = rb_ivar_get(context, id_ivar_resource_limits);
-    rb_funcall(resource_limits, id_increment_render_score, 1, INT2NUM(body->render_score));
+static void vm_free(void *ptr)
+{
+    vm_t *vm = ptr;
+    c_buffer_free(&vm->stack);
+    xfree(vm);
+}
 
-    bool is_captured = rb_ivar_get(resource_limits, id_ivar_last_capture_length) != Qnil;;
-    long render_length_limit = LONG_MAX;
+static size_t vm_memsize(const void *ptr)
+{
+    const vm_t *vm = ptr;
+    if (!vm) return 0;
+    return sizeof(vm_t) + c_buffer_capacity(&vm->stack);
+}
 
-    if (!is_captured) {
-        VALUE render_length_limit_num = rb_ivar_get(resource_limits, id_ivar_render_length_limit);
-        if (render_length_limit_num != Qnil)
-            render_length_limit = NUM2LONG(render_length_limit_num);
+const rb_data_type_t vm_data_type = {
+    "liquid_vm",
+    { vm_mark, vm_free, vm_memsize, },
+    NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static VALUE vm_internal_new(VALUE context)
+{
+    vm_t *vm;
+    VALUE obj = TypedData_Make_Struct(cLiquidCVM, vm_t, &vm_data_type, vm);
+    vm->stack = c_buffer_init();
+
+    vm->strainer = rb_funcall(context, id_strainer, 0);;
+    Check_Type(vm->strainer, T_OBJECT);
+
+    vm->filter_methods = rb_funcall(RBASIC_CLASS(vm->strainer), id_filter_methods_hash, 0);;
+    Check_Type(vm->filter_methods, T_HASH);
+
+    vm->interrupts = rb_ivar_get(context, id_ivar_interrupts);
+    Check_Type(vm->interrupts, T_ARRAY);
+
+    vm->strict_filters = RTEST(rb_funcall(context, id_strict_filters, 0));
+    vm->global_filter = rb_funcall(context, id_global_filter, 0);
+    vm->resource_limits = rb_ivar_get(context, id_ivar_resource_limits);;
+    return obj;
+}
+
+static vm_t *vm_from_context(VALUE context)
+{
+    VALUE vm_obj = rb_attr_get(context, id_vm);
+    if (vm_obj == Qnil) {
+        vm_obj = vm_internal_new(context);
+        rb_ivar_set(context, id_vm, vm_obj);
+    }
+    // instance variable is hidden from C so should be safe to unwrap it without type checking
+    return DATA_PTR(vm_obj);
+}
+
+static void write_fixnum(VALUE output, VALUE fixnum)
+{
+    long long number = RB_NUM2LL(fixnum);
+    int write_length = snprintf(NULL, 0, "%lld", number);
+    long old_size = RSTRING_LEN(output);
+    long new_size = old_size + write_length;
+    long capacity = rb_str_capacity(output);
+
+    if (new_size > capacity) {
+        do {
+            capacity *= 2;
```

Comment:
This is unlikely to be the last 10 bytes, so doubling is meant to amortize the cost of resizing the output string.

> Lets say the output buffer is 10KB, doubling for the sake of adding 10 bytes?
> 
> Maybe this could rewritten to use [`rb_str_buf_cat_ascii`](https://github.com/Shopify/ruby/blob/master/string.c#L3017), and let Ruby handle the resizing?

You would actually have approximately the same behaviour.  I created https://github.com/dylanahsmith/ruby-string-buffer a long time ago to make it easier to inspect the capacity.

```
irb> str = "1" * (10 * 1024);
irb> StringBuffer.capacity(str)
=> 10240
irb> str << "2";
irb> StringBuffer.capacity(str)
=> 20481
```

which actually allocates 1 extra byte, which keeps the allocation from being page aligned.

> Writing the fixnum to a `char number_str[11]` on the stack.

We would still have to resize the output buffer to copy it from the stack to the output buffer, so this approach avoids the copy step.

> Or replace the whole thing w/ `rb_str_buf_append(output, rb_fix2str(fixnum));`?

That would trade-off performance for simplicity, since it would allocate a new string object to write the result to as well as having to copy the result to the output buffer.  The primary reason for writing `write_fixnum` was to avoid that string allocation overhead.